internal/sourceanalysis/rust.go

```diff
@@ -0,0 +1,283 @@
+	"github.com/google/osv-scanner/internal/thirdparty/ar"
+	"github.com/google/osv-scanner/pkg/models"
+	"github.com/google/osv-scanner/pkg/reporter"
+	"github.com/ianlancetaylor/demangle"
+)
+
+const (
+	RustFlagsEnv     = "RUSTFLAGS=-C opt-level=3 -C debuginfo=1"
+	RustLibExtension = ".rcgu.o/"
+)
+
+func rustAnalysis(r reporter.Reporter, pkgs []models.PackageVulns, source models.SourceInfo) {
+	binaryPaths, err := rustBuildSource(r, source)
+	if err != nil {
+		r.PrintError(fmt.Sprintf("failed to build cargo/rust project from source: %s", err))
+		return
+	}
+
+	// This map stores 3 states for each vuln ID
+	// - There is function level vuln info, but it **wasn't** called   (false)
+	// - There is function level vuln info, and it **is** called    (true)
+	// - There is **no** functional level vuln info, so we don't know whether it is called (doesn't exist)
+	isCalledVulnMap := map[string]bool{}
+
+	for _, path := range binaryPaths {
+		if strings.HasSuffix(path, ".rlib") {
+			// Is a library, so need an extra step to extract the object binary file before passing to parseDWARFData
+			objFilePath, err := extractRlibArchive(path)
+			if err != nil {
+				r.PrintError(fmt.Sprintf("failed to analyse '%s': %s", path, err))
+				continue
+			}
+			// TODO: Do we need to care about error on deletion here?
+			defer os.Remove(objFilePath)
+		}
+		calls, err := parseDWARFData(r, path)
+		if err != nil {
+			r.PrintError(fmt.Sprintf("failed to analyse '%s': %s", path, err))
+			continue
+		}
+
+		for _, pv := range pkgs {
+			for _, v := range pv.Vulnerabilities {
+				for _, a := range v.Affected {
+					// Example of RUSTSEC function level information:
+					//
+					// "affects": {
+					//     "os": [],
+					//     "functions": [
+					//         "smallvec::SmallVec::grow"
+					//     ],
+					//     "arch": []
+					// }
+					ecosystemAffects, ok := a.EcosystemSpecific["affects"].(map[string]interface{})
+					if !ok {
+						continue
+					}
+					affectedFunctions, ok := ecosystemAffects["functions"].([]interface{})
+					if !ok {
+						continue
+					}
+					for _, f := range affectedFunctions {
+						if funcName, ok := f.(string); ok {
+							_, called := calls[funcName]
+							// Once one advisory marks this vuln as called, always mark as called
+							isCalledVulnMap[v.ID] = isCalledVulnMap[v.ID] || called
+						}
+					}
+				}
+			}
+		}
+
+		for _, pv := range pkgs {
+			for groupIdx := range pv.Groups {
+				for _, vulnID := range pv.Groups[groupIdx].IDs {
+					analysis := &pv.Groups[groupIdx].ExperimentalAnalysis
+					if *analysis == nil {
+						*analysis = make(map[string]models.AnalysisInfo)
+					}
+
+					called, hasFuncInfo := isCalledVulnMap[vulnID]
+					if hasFuncInfo {
+						(*analysis)[vulnID] = models.AnalysisInfo{
+							Called: called,
+						}
+					}
+				}
+			}
+		}
+	}
+}
+
+func parseDWARFData(_ reporter.Reporter, binaryPath string) (map[string]struct{}, error) {
```

Comment:
nit: can we make the name more descriptive. e.g. `functionsFromDWARF` 

Also, why do we need the `reporter.Reporter` argument if it's not used?